File: clients/ios/Sample/UITests/RemovableE2ETest.swift
import XCTest

/// On-device OP_REMOVE_NODE (the `remove` sentinel). Navigate to
/// /native/removable and tap "Hide": the banner's dynamic returns `remove`, the
/// server diffs to op 4, and the client drops the node rather than crash. The
/// Android analogue is RemovableE2ETest.
final class RemovableE2ETest: XCTestCase {
    private var app: XCUIApplication!

    override func setUpWithError() throws {
        continueAfterFailure = false
        app = launchSample()
    }

    func testDropsNodeOnTheRemoveSentinel() {
        tapButton(app, "Removable")
        waitForText(app, "Banner!")

        tapButton(app, "Hide")
        waitForNoText(app, "Banner!")
    }
}

File: clients/android/sample/src/androidTest/kotlin/dev/arizona/sample/RemovableE2ETest.kt
package dev.arizona.sample

import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import org.junit.Rule
import org.junit.Test

/**
 * On-device OP_REMOVE_NODE (the `remove` sentinel). Navigate to /native/removable
 * and tap "Hide": the banner's text dynamic returns `remove`, the server diffs to
 * op 4, and the client must drop the node rather than crash. The Playwright
 * analogue is e2e/native/removable.spec.js.
 *
 * Requires the Arizona test server reachable at localhost:4040 (adb reverse
 * tcp:4040 tcp:4040). Run with: ./gradlew :sample:connectedCheck
 */
class RemovableE2ETest {
    @get:Rule
    val rule = createAndroidComposeRule<MainActivity>()

    @Test
    fun dropsNodeOnTheRemoveSentinel() {
        rule.waitForText("Removable")
        rule.onNodeWithText("Removable").performClick()
        rule.waitForText("Banner!")

        rule.onNodeWithText("Hide").performClick()
        rule.waitForNoText("Banner!")
    }
}
